server/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -476,24 +477,24 @@ public Version getAsVersion(String setting, Version defaultVersion) throws Setti
      * @return  The direct keys of this settings
      */
     public Set<String> names() {
-        synchronized (firstLevelNames) {
-            if (firstLevelNames.get() == null) {
-                Stream<String> stream = settings.keySet().stream();
-                if (secureSettings != null) {
-                    stream = Stream.concat(stream, secureSettings.getSettingNames().stream());
-                }
-                Set<String> names = stream.map(k -> {
-                    int i = k.indexOf('.');
-                    if (i < 0) {
-                        return k;
-                    } else {
-                        return k.substring(0, i);
-                    }
-                }).collect(Collectors.toSet());
-                firstLevelNames.set(Collections.unmodifiableSet(names));
+        final Set<String> names = firstLevelNames;
```

Comment:
I think removing the synchronized would work. The only possibility of breaking the names creation with multiple threads is if someone mutated settings or secure settings while we were iterating on them. However, in that case the lock on firstLevelNames wouldn't help us much.

If nobody mutates settings or secure settings after we've taken a reference to them in the constructor (the original code doesn't guard against it anyway), the only behaviour change that this can introduce is if anyone consuming the resulting Set is doing == instead of .equals on the set or on the Strings in the set. 

I think using == would be considered a bug on the consumer side?

Reply:
Similar pattern exists inside `PrefixedSecureSettings.getSettingNames` below. Does it make sense to remove the synchronized block in that method too?

Reply:
>  if someone mutated settings

This shouldn't be possible since Settings are immutable. We use Settings.Builder to add settings, then construct an immutable instance.

> I think using == would be considered a bug on the consumer side?

Yes, nothing here should be relying on reference equality, the key name methods don't state anything in their javadocs about expecting the same object returned.